transcendence-app/src/chat/chat.service.ts

```diff
@@ -175,4 +179,46 @@ export class ChatService {
     }
     return this.chatroomMemberService.addChatroomMember(chatroomId, userId);
   }
+
+  async updateChatroomMember(
+    authUser: User,
+    chatroomId: string,
+    userId: string,
+    updateChatroomMemberDto: UpdateChatroomMemberDto,
+  ): Promise<ChatroomMember | null> {
+    const authChatroomMember =
+      await this.chatroomMemberRepository.getByIdWithUser(
+        chatroomId,
+        authUser.id,
+      );
+    const chatroomMember = await this.chatroomMemberRepository.getByIdWithUser(
+      chatroomId,
+      userId,
+    );
+    if (!authChatroomMember || !chatroomMember) {
+      throw new NotFoundException();
+    }
+    if (chatroomMember.owner) {
+      throw new ForbiddenException();
+    }
+    if (authChatroomMember.owner) {
+      return this.chatroomMemberService.updateById(
+        chatroomId,
+        userId,
+        updateChatroomMemberDto,
+      );
+    }
+    if (
+      !authChatroomMember.admin ||
+      chatroomMember.admin ||
+      updateChatroomMemberDto.admin !== undefined
+    ) {
+      throw new ForbiddenException();
+    }
```

Comment:
A banned admin cannot update itself, and also can still update non owners, even if the admin is banned. Does this make sense?

Reply:
Good catch! What do we think is a better option to solve this?

1. Add a check constraint on the `ChatMember` table to disallow a chatroom member that is admin and banned/muted. For example:

```sql
    CHECK (
      NOT (
        (
          admin IS true
          AND banned IS true
        )
        OR (
          admin IS true
          AND muted IS true
        )
      )
    )
```
2. Add a check to this condition.

Reply:
I don't think this is a good solution:

> When owner of a channel, user can kick, ban, mute other users and administrators When administrator of a channel, user can kick, ban mute other users, but not owners

Reply:
I see. With this solution (table constraint check), the owner has to remove the admin rights before banning/muting an admin.

Then it is better to add some condition to check that the authenticated user is an admin and it's not banned or muted.

Reply:
BTW. If the admin is not banned but muted, should we allow them to ban/mute other users?

Reply:
Updated. I've added the conditions to this `if` statement.